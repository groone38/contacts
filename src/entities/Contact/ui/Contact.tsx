import React from "react";

import { Link } from "react-router-dom";

import classes from "./Contact.module.scss";

interface ContactProps {
  id: string;
  name: string;
}

const Contact = ({ id, name }: ContactProps) => {
  return (
    <Link to={`/${id}`} className={classes.contact}>
      <p>{name}</p>
    </Link>
  );
};

export default Contact;
