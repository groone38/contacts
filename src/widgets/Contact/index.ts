import { contacts } from "src/features";

export { contacts };
