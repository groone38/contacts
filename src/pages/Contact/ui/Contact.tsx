import React, { useState, useEffect } from "react";
import classes from "./Contact.module.scss";
import { useNavigate, useParams } from "react-router-dom";
import {
  deleteContact,
  getOneContact,
} from "src/features/model/reducers/ContactsSlice";
import { useAppDispatch, useAppSelector } from "src/app/providers/store";
import Delete from "../../../shared/img/delete.svg";
import Edit from "../../../shared/img/edit.svg";
import { SubmitHandler, useForm } from "react-hook-form";
import { Input } from "src/shared/Input";
import { FaCheck, FaRegCircleXmark } from "react-icons/fa6";
import { Button } from "src/shared/ui/Button";

interface ModalFormData {
  email: string;
  first_name: string;
  last_name: string;
  tel: null | number;
  company: string;
  about: string;
}

const Contact = () => {
  const [edit, setEdit] = useState<boolean>(false);
  const data = useAppSelector(
    (state) => state.contacts.contact
  ) as ModalFormData;
  const loading = useAppSelector((state) => state.contacts.loading);

  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm<ModalFormData>({
    mode: "all",
  });
  const { id } = useParams();
  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(getOneContact(id));
    const defaultValues: ModalFormData = {
      first_name: "",
      company: "",
      about: "",
      email: "",
      last_name: "",
      tel: null,
    };
    defaultValues.first_name = data?.first_name;
    defaultValues.about = data?.about;
    defaultValues.company = data?.company;
    defaultValues.email = data?.email;
    defaultValues.last_name = data?.last_name;
    defaultValues.tel = data?.tel;
    reset({ ...defaultValues });
  }, [edit]);

  const remove = async () => {
    dispatch(deleteContact(id)).then(() => navigate("/"));
  };

  const onSubmit: SubmitHandler<ModalFormData> = async (data) => {
    console.log(data);
    setEdit(false);
  };

  return (
    <>
      {loading && <h1>Loading...</h1>}
      <div className={classes.contact}>
        <form
          onSubmit={handleSubmit(onSubmit)}
          className={classes.contact__block}
        >
          <div>
            {edit ? (
              <Input
                id="first_name"
                name="first_name"
                placeholder="Enter you first name"
                register={register}
                required="Поле First name не может быть пустым!"
                type="text"
                label="First name"
                errors={errors?.first_name?.message}
              />
            ) : (
              <>
                <strong>Name: </strong>
                <span>{data?.first_name}</span>
              </>
            )}
          </div>

          <div>
            {edit ? (
              <Input
                id="last_name"
                name="last_name"
                placeholder="Enter you last name"
                register={register}
                required="Поле Last name не может быть пустым!"
                type="text"
                label="Last name"
                errors={errors?.last_name?.message}
              />
            ) : (
              <>
                <strong>Last name: </strong>
                <span>{data?.last_name}</span>
              </>
            )}
          </div>
          <div>
            {edit ? (
              // <div>
              <Input
                id="email"
                name="email"
                placeholder="Enter you email"
                register={register}
                required="Поле Email не может быть пустым!"
                type="email"
                label="Email"
                errors={errors?.email?.message}
              />
            ) : (
              /* {errors && <span>{errors?.email?.message}</span>} */
              /* </div> */
              <>
                <strong>Email: </strong>
                <span>{data?.email}</span>
              </>
            )}
          </div>
          <div>
            {edit ? (
              // <div>
              <Input
                id="company"
                name="company"
                placeholder="Enter you company"
                register={register}
                required="Поле Company не может быть пустым!"
                type="text"
                label="Company"
                errors={errors?.company?.message}
              />
            ) : (
              /* {errors && <span>{errors?.company?.message}</span>}
              </div> */
              <>
                <strong>Company: </strong>
                <span>{data?.company}</span>
              </>
            )}
          </div>
          <div>
            {edit ? (
              // <div>
              <Input
                id="tel"
                name="tel"
                placeholder="Enter you tel"
                register={register}
                required="Поле Tel не может быть пустым!"
                type="number"
                label="Tel"
                errors={errors?.tel?.message}
              />
            ) : (
              /* {errors && <span>{errors?.tel?.message}</span>} */
              /* </div> */
              <>
                <strong>Tel: </strong>
                <span>{data?.tel}</span>
              </>
            )}
          </div>
          {/* <ContactInput edit={edit} id="about"
                  name="about"
                  placeholder="Enter you about"
                  register={register}
                  required="Поле About не может быть пустым!"
                  type="text"
                  label="About"
                  errors={errors?.tel?.message}
                  /> */}
          <div>
            {edit ? (
              // <div>
              <Input
                id="about"
                name="about"
                placeholder="Enter you about"
                register={register}
                required="Поле About не может быть пустым!"
                type="text"
                label="About"
                errors={errors?.about?.message}
              />
            ) : (
              /* {errors && <span>{errors?.about?.message}</span>} */
              // </div>
              <>
                <strong>About: </strong>
                <span>{data?.about}</span>
              </>
            )}
          </div>
          <div className={classes.contact__btn}>
            {edit ? (
              <>
                <button type="submit">
                  <FaCheck fill="white" />
                </button>
                <button onClick={() => setEdit(false)}>
                  <FaRegCircleXmark fill="white" />
                </button>
              </>
            ) : (
              <>
                <Button onclick={() => setEdit(true)} type="button">
                  <img src={Edit} alt="edit" />
                </Button>
                <Button onclick={remove} type="button">
                  <img src={Delete} alt="delete" />
                </Button>
              </>
            )}
          </div>
        </form>
      </div>
    </>
  );
};

export default Contact;
