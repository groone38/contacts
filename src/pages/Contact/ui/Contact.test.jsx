import { render, screen } from "@testing-library/react";
import * as reduxHooks from "react-redux";
import Contact from "./Contact";
import { BrowserRouter } from "react-router-dom";
import "@testing-library/jest-dom";

jest.mock("react-redux");

const mockDispatch = jest.spyOn(reduxHooks, "useDispatch");
const mockSelector = jest.spyOn(reduxHooks, "useSelector");

const user = {
  first_name: "Andrey",
  company: "LisDev",
  about: "Frontend dev",
  email: "groone38@yandex.ru",
  last_name: "Lisin",
  tel: 89054437359,
};

describe("Contact component", () => {
  it("render", async () => {
    mockDispatch.mockReturnValue(jest.fn());
    mockSelector.mockReturnValue([]);
    const contact = render(
      <BrowserRouter>
        <Contact />
      </BrowserRouter>
    );
    expect(contact).toMatchSnapshot();
  });
  it("show Contact", async () => {
    mockDispatch.mockReturnValue(jest.fn());
    mockSelector.mockReturnValue(user);
    render(
      <BrowserRouter>
        <Contact />
      </BrowserRouter>
    );

    expect(screen.getByText("Frontend dev")).toBeInTheDocument();
  });
  it("contact snapchot", () => {
    mockDispatch.mockReturnValue(jest.fn());
    mockSelector.mockReturnValue(user);
    const contact = render(
      <BrowserRouter>
        <Contact />
      </BrowserRouter>
    );
    expect(contact).toMatchSnapshot();
  });
});
