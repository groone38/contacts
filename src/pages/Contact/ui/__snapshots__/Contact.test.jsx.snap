// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact component contact snapchot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="overlay"
      >
        <div
          class="lds_spinner"
        >
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
      <div
        class="contact"
      >
        <form
          class="contact__block"
        >
          <div
            class="info"
          >
            <strong>
              First name
              : 
            </strong>
            <span>
              Andrey
            </span>
          </div>
          <div
            class="info"
          >
            <strong>
              Last name
              : 
            </strong>
            <span>
              Lisin
            </span>
          </div>
          <div
            class="info"
          >
            <strong>
              Email
              : 
            </strong>
            <span>
              groone38@yandex.ru
            </span>
          </div>
          <div
            class="info"
          >
            <strong>
              Company
              : 
            </strong>
            <span>
              LisDev
            </span>
          </div>
          <div
            class="info"
          >
            <strong>
              Tel
              : 
            </strong>
            <span>
              89054437359
            </span>
          </div>
          <div
            class="info"
          >
            <strong>
              About
              : 
            </strong>
            <span>
              Frontend dev
            </span>
          </div>
          <div
            class="contact__btn"
          >
            <button
              class="btn"
              data-testid="btn"
              type="button"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M410.3 231l11.3-11.3-33.9-33.9-62.1-62.1L291.7 89.8l-11.3 11.3-22.6 22.6L58.6 322.9c-10.4 10.4-18 23.3-22.2 37.4L1 480.7c-2.5 8.4-.2 17.5 6.1 23.7s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L387.7 253.7 410.3 231zM160 399.4l-9.1 22.7c-4 3.1-8.5 5.4-13.3 6.9L59.4 452l23-78.1c1.4-4.9 3.8-9.4 6.9-13.3l22.7-9.1v32c0 8.8 7.2 16 16 16h32zM362.7 18.7L348.3 33.2 325.7 55.8 314.3 67.1l33.9 33.9 62.1 62.1 33.9 33.9 11.3-11.3 22.6-22.6 14.5-14.5c25-25 25-65.5 0-90.5L453.3 18.7c-25-25-65.5-25-90.5 0zm-47.4 168l-144 144c-6.2 6.2-16.4 6.2-22.6 0s-6.2-16.4 0-22.6l144-144c6.2-6.2 16.4-6.2 22.6 0s6.2 16.4 0 22.6z"
                />
              </svg>
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="overlay"
    >
      <div
        class="lds_spinner"
      >
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
      </div>
    </div>
    <div
      class="contact"
    >
      <form
        class="contact__block"
      >
        <div
          class="info"
        >
          <strong>
            First name
            : 
          </strong>
          <span>
            Andrey
          </span>
        </div>
        <div
          class="info"
        >
          <strong>
            Last name
            : 
          </strong>
          <span>
            Lisin
          </span>
        </div>
        <div
          class="info"
        >
          <strong>
            Email
            : 
          </strong>
          <span>
            groone38@yandex.ru
          </span>
        </div>
        <div
          class="info"
        >
          <strong>
            Company
            : 
          </strong>
          <span>
            LisDev
          </span>
        </div>
        <div
          class="info"
        >
          <strong>
            Tel
            : 
          </strong>
          <span>
            89054437359
          </span>
        </div>
        <div
          class="info"
        >
          <strong>
            About
            : 
          </strong>
          <span>
            Frontend dev
          </span>
        </div>
        <div
          class="contact__btn"
        >
          <button
            class="btn"
            data-testid="btn"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M410.3 231l11.3-11.3-33.9-33.9-62.1-62.1L291.7 89.8l-11.3 11.3-22.6 22.6L58.6 322.9c-10.4 10.4-18 23.3-22.2 37.4L1 480.7c-2.5 8.4-.2 17.5 6.1 23.7s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L387.7 253.7 410.3 231zM160 399.4l-9.1 22.7c-4 3.1-8.5 5.4-13.3 6.9L59.4 452l23-78.1c1.4-4.9 3.8-9.4 6.9-13.3l22.7-9.1v32c0 8.8 7.2 16 16 16h32zM362.7 18.7L348.3 33.2 325.7 55.8 314.3 67.1l33.9 33.9 62.1 62.1 33.9 33.9 11.3-11.3 22.6-22.6 14.5-14.5c25-25 25-65.5 0-90.5L453.3 18.7c-25-25-65.5-25-90.5 0zm-47.4 168l-144 144c-6.2 6.2-16.4 6.2-22.6 0s-6.2-16.4 0-22.6l144-144c6.2-6.2 16.4-6.2 22.6 0s6.2 16.4 0 22.6z"
              />
            </svg>
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Contact component render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="overlay"
      >
        <div
          class="lds_spinner"
        >
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
      <div
        class="contact"
      >
        <form
          class="contact__block"
        >
          <div
            class="info"
          >
            <strong>
              First name
              : 
            </strong>
            <span />
          </div>
          <div
            class="info"
          >
            <strong>
              Last name
              : 
            </strong>
            <span />
          </div>
          <div
            class="info"
          >
            <strong>
              Email
              : 
            </strong>
            <span />
          </div>
          <div
            class="info"
          >
            <strong>
              Company
              : 
            </strong>
            <span />
          </div>
          <div
            class="info"
          >
            <strong>
              Tel
              : 
            </strong>
            <span />
          </div>
          <div
            class="info"
          >
            <strong>
              About
              : 
            </strong>
            <span />
          </div>
          <div
            class="contact__btn"
          >
            <button
              class="btn"
              data-testid="btn"
              type="button"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M410.3 231l11.3-11.3-33.9-33.9-62.1-62.1L291.7 89.8l-11.3 11.3-22.6 22.6L58.6 322.9c-10.4 10.4-18 23.3-22.2 37.4L1 480.7c-2.5 8.4-.2 17.5 6.1 23.7s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L387.7 253.7 410.3 231zM160 399.4l-9.1 22.7c-4 3.1-8.5 5.4-13.3 6.9L59.4 452l23-78.1c1.4-4.9 3.8-9.4 6.9-13.3l22.7-9.1v32c0 8.8 7.2 16 16 16h32zM362.7 18.7L348.3 33.2 325.7 55.8 314.3 67.1l33.9 33.9 62.1 62.1 33.9 33.9 11.3-11.3 22.6-22.6 14.5-14.5c25-25 25-65.5 0-90.5L453.3 18.7c-25-25-65.5-25-90.5 0zm-47.4 168l-144 144c-6.2 6.2-16.4 6.2-22.6 0s-6.2-16.4 0-22.6l144-144c6.2-6.2 16.4-6.2 22.6 0s6.2 16.4 0 22.6z"
                />
              </svg>
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="overlay"
    >
      <div
        class="lds_spinner"
      >
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
      </div>
    </div>
    <div
      class="contact"
    >
      <form
        class="contact__block"
      >
        <div
          class="info"
        >
          <strong>
            First name
            : 
          </strong>
          <span />
        </div>
        <div
          class="info"
        >
          <strong>
            Last name
            : 
          </strong>
          <span />
        </div>
        <div
          class="info"
        >
          <strong>
            Email
            : 
          </strong>
          <span />
        </div>
        <div
          class="info"
        >
          <strong>
            Company
            : 
          </strong>
          <span />
        </div>
        <div
          class="info"
        >
          <strong>
            Tel
            : 
          </strong>
          <span />
        </div>
        <div
          class="info"
        >
          <strong>
            About
            : 
          </strong>
          <span />
        </div>
        <div
          class="contact__btn"
        >
          <button
            class="btn"
            data-testid="btn"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M410.3 231l11.3-11.3-33.9-33.9-62.1-62.1L291.7 89.8l-11.3 11.3-22.6 22.6L58.6 322.9c-10.4 10.4-18 23.3-22.2 37.4L1 480.7c-2.5 8.4-.2 17.5 6.1 23.7s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L387.7 253.7 410.3 231zM160 399.4l-9.1 22.7c-4 3.1-8.5 5.4-13.3 6.9L59.4 452l23-78.1c1.4-4.9 3.8-9.4 6.9-13.3l22.7-9.1v32c0 8.8 7.2 16 16 16h32zM362.7 18.7L348.3 33.2 325.7 55.8 314.3 67.1l33.9 33.9 62.1 62.1 33.9 33.9 11.3-11.3 22.6-22.6 14.5-14.5c25-25 25-65.5 0-90.5L453.3 18.7c-25-25-65.5-25-90.5 0zm-47.4 168l-144 144c-6.2 6.2-16.4 6.2-22.6 0s-6.2-16.4 0-22.6l144-144c6.2-6.2 16.4-6.2 22.6 0s6.2 16.4 0 22.6z"
              />
            </svg>
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
